### PR TITLE
fix memory leak

### PR DESCRIPTION
**Pull request by stanleyljl**
Re: cocos-creator/3d-tasks#

Changelog:
 * fix some memory leaks

### CODE REVIEW

File: native/cocos/application/ApplicationManager.cpp [hunk 1]
```diff
@@ -50,9 +56,14 @@ ApplicationManager::ApplicationPtr ApplicationManager::getCurrentAppSafe() const
 }
 } // namespace cc
 
-//
+void cocos_close() { // NOLINT(readability-identifier-naming)
+    // Called in the platform layer, because the platform layer is isolated from the application layer
+    // It is the platform layer to drive applications and reclaim resources.
+    cc::ApplicationManager::getInstance()->closeAllApplications();
```
Comment (minggo) [inline native/cocos/application/ApplicationManager.cpp:62]: Can do it in `cocos_destory()`?

File: native/cocos/engine/Engine.cpp [hunk 2]
```diff
@@ -51,6 +51,8 @@
 #include "cocos/network/HttpClient.h"
 #include "core/Root.h"
 #include "core/assets/FreeTypeFont.h"
+#include "editor-support/dragonbones-creator-support/ArmatureCacheMgr.h"
+#include "editor-support/spine-creator-support/SkeletonCacheMgr.h"
```
Comment (minggo) [inline native/cocos/engine/Engine.cpp]: spine and dragonbones may not be included into engine. Need to use corresponding macro to include these codes.

File: native/cocos/renderer/gfx-gles2/GLES2Wrangler.cpp [hunk 3]
```diff
@@ -83,3 +114,12 @@ bool gles2wInit() {
 
     return true;
 }
+
+
+bool gles2wExit() {
+    if (!gles2wClose()) {
```
Comment (minggo) [inline native/cocos/renderer/gfx-gles2/GLES2Wrangler.cpp]: Why need gles2wExit? Can just use gles2Close?

Comment (minggo) [reply native/cocos/renderer/gfx-gles2/GLES2Wrangler.cpp]: I got it, there is a init function.

File: native/cocos/scene/Camera.cpp [hunk 4]
```diff
@@ -110,7 +111,6 @@ void Camera::destroy() {
         _window = nullptr;
     }
     _name.clear();
-    _geometryRenderer->destroy();
```
Comment (minggo) [inline native/cocos/scene/Camera.cpp:113]: Why not destroy it in destroy function?

File: native/cocos/application/ApplicationManager.cpp (hunk 1 above)
Comment (stanleyljl) [reply native/cocos/application/ApplicationManager.cpp:62]: The entire exit process needs to be refactored, so here is the simple treatment

File: native/cocos/engine/Engine.cpp (hunk 2 above)
Comment (stanleyljl) [reply native/cocos/engine/Engine.cpp]: Where is the corresponding macro definition？

File: native/cocos/renderer/gfx-gles2/GLES2Wrangler.cpp (hunk 3 above)
Comment (dumganhar) [inline native/cocos/renderer/gfx-gles2/GLES2Wrangler.cpp]: What's the difference between gles2wExit & gles2wClose?

File: native/cocos/engine/Engine.cpp [hunk 5]
```diff
@@ -154,29 +157,18 @@ int Engine::restart() {
 }
 
 void Engine::close() { // NOLINT
-    if (cc::EventDispatcher::initialized()) {
-        cc::EventDispatcher::dispatchCloseEvent();
-    }
```
Comment (minggo) [inline native/cocos/engine/Engine.cpp:159]: Why not dispatch close event?

File: native/cocos/scene/Camera.cpp (hunk 4 above)
Comment (stanleyljl) [reply native/cocos/scene/Camera.cpp:113]: The destroy function of Camera is never called.

File: native/cocos/application/ApplicationManager.cpp (hunk 1 above)
Comment (minggo) [reply native/cocos/application/ApplicationManager.cpp:62]: What's the problem of exit process?

File: native/cocos/scene/Camera.cpp (hunk 4 above)
Comment (dumganhar) [inline native/cocos/scene/Camera.cpp:113]: Is it better to keep this line and  set _geometryRenderer to nullptr?
```c++
Camera::~Camera() {
    CC_SAFE_DESTROY_NULL(_geometryRenderer); // if _geometryRenderer is nullptr, then do nothing.
    _frustum->release();
}

void Camera::destroy() {
    if (_window) {
        _window->detachCamera(this);
        _window = nullptr;
    }
    _name.clear();

    CC_SAFE_DESTROY_NULL(_geometryRenderer); // If _geometryRenderer is not nullptr, invoke destroy method and set it to nullptr.
```

File: native/cocos/engine/Engine.cpp (hunk 2 above)
Comment (minggo) [reply native/cocos/engine/Engine.cpp]: You can find them in CMakeLists.txt, such as CC_USE_SPINE.

File: native/cocos/scene/Camera.cpp (hunk 4 above)
Comment (minggo) [reply native/cocos/scene/Camera.cpp:113]: If so, then the logic is error. We should fix the logic. Currently, destructor invoke destroy function is better.

File: native/cocos/engine/Engine.cpp [hunk 6]
```diff
@@ -154,29 +157,18 @@ int Engine::restart() {
 }
 
 void Engine::close() { // NOLINT
-    if (cc::EventDispatcher::initialized()) {
-        cc::EventDispatcher::dispatchCloseEvent();
```
Comment (dumganhar) [inline native/cocos/engine/Engine.cpp:158]: Why delete this line ?
It's used for dispatching `onClose` event to JS.

File: native/cocos/application/ApplicationManager.cpp (hunk 1 above)
Comment (stanleyljl) [reply native/cocos/application/ApplicationManager.cpp:62]: The initialization process and exit process of the entire engine need to be carefully designed, and it is still confusing now.

File: native/cocos/renderer/gfx-gles2/GLES2Wrangler.cpp (hunk 3 above)
Comment (stanleyljl) [reply native/cocos/renderer/gfx-gles2/GLES2Wrangler.cpp]: The exit process mimics the initialization process：gles2wInit->gles2wOpen

File: native/cocos/application/ApplicationManager.cpp (hunk 1 above)
Comment (qiuguohua) [reply native/cocos/application/ApplicationManager.cpp:62]: Yes, the whole exit needs to be refactored. Calling cocos_close is to call engine.close(). Here, exit(0) will be called; basically, the logic behind will not be called. This was originally done this way, and I've kept the operation.

File: native/cocos/engine/Engine.cpp (hunk 5 above)
Comment (stanleyljl) [reply native/cocos/engine/Engine.cpp:159]: dispatchCloseEvent is called before somewhere.

File: native/cocos/scene/Camera.cpp (hunk 4 above)
Comment (stanleyljl) [reply native/cocos/scene/Camera.cpp:113]: 1

Comment (stanleyljl) [reply native/cocos/scene/Camera.cpp:113]: 1

File: native/cocos/engine/Engine.cpp (hunk 6 above)
Comment (stanleyljl) [reply native/cocos/engine/Engine.cpp:158]: It is called before

File: native/cocos/scene/Camera.cpp (hunk 4 above)
Comment (stanleyljl) [reply native/cocos/scene/Camera.cpp:113]: 1

File: native/cocos/application/ApplicationManager.cpp (hunk 1 above)
Comment (minggo) [reply native/cocos/application/ApplicationManager.cpp:62]: > The initialization process and exit process of the entire engine need to be carefully designed, and it is still confusing now.

Ok. Then it is better not export more functions.

File: native/cocos/engine/Engine.cpp (hunk 5 above)
Comment (minggo) [reply native/cocos/engine/Engine.cpp:159]: @qiuguohua please double check it.

**Review by dumganhar — APPROVED** [review]

File: native/cocos/application/ApplicationManager.cpp [hunk 7]
```diff
@@ -33,6 +33,12 @@ ApplicationManager *ApplicationManager::getInstance() {
     return &mgr;
 }
 
+void ApplicationManager::closeAllApplications() {
+    for (auto &app : _apps) {
```
Comment (dumganhar) [inline native/cocos/application/ApplicationManager.cpp:37]: Use `for (auto* app : _apps)`
`auto*` for pointer rather than `auto&`

Comment (stanleyljl) [reply native/cocos/application/ApplicationManager.cpp:37]: ApplicationPtr is not a raw pointer. The code will compile with errors

File: native/cocos/engine/Engine.cpp (hunk 5 above)
Comment (qiuguohua) [reply native/cocos/engine/Engine.cpp:159]: > dispatchCloseEvent is called before somewhere.

inside onclose? The timing of these two calls is different. 'close' is an active close, and exit will be called inside, but onclose will not be called?

'onclose' is called when the window is closed

File: native/cocos/application/ApplicationManager.cpp (hunk 7 above)
Comment (dumganhar) [reply native/cocos/application/ApplicationManager.cpp:37]: Got it.

File: native/cocos/engine/Engine.cpp (hunk 5 above)
Comment (stanleyljl) [reply native/cocos/engine/Engine.cpp:159]: It is called in Engine::dispatchWindowEvent, which is before Engine::close

**Review by qiuguohua — APPROVED** [review]

**Review by minggo — APPROVED** [review]

File: native/cocos/application/ApplicationManager.cpp (hunk 1 above)
Comment (minggo) [reply native/cocos/application/ApplicationManager.cpp:62]: @qiuguohua please create an issue to refactor the init and exit process.

**Pull request merged**